posts/2022/01/ipython-8.0-lessons-learned-maintaining-software.md

```diff
@@ -0,0 +1,303 @@
+ - it's explicit.
+ - it's easier to search for. 
+ - it's easier to remove.
+
+
+ For example IPython used to contain
+
+```
+try:
+    from numpy.testing import KnownFailure, knownfailureif
+except ImportError:
+    from ._decorators import knownfailureif
+    try:
+        from ._numpy_testing_noseclasses import KnownFailure
+    except ImportError:
+        pass
+```
+
+
+Which adds a fallback for [numpy version older than numpy 1.3 from 2008](https://github.com/numpy/numpy/commit/ba9a02dcb2c3ca635076a75cc9eb0f406e00ceed).
+It tooks me ~30 minutes to find this informations, which could have been seconds would the author (which could have been
+me) had checked version.  A proper version check would also had this code removed years ago. 
+
+While IPython has few dependencies beyond Python, we do so with Python version itself, and always compare with
+`sys.version_info` made it straightforward to find all dead code once we bumped or minimal version to 3.8+.
+
+This has domino effects in IPython 8.0 as we were going to great length to support top level async which is since 3.8
+native to Python. Many simplification leading do even more down the line up to complete method and class suppressions.
+
+
+# Don't be cheap on `DeprecationWarning`s
+
+Warnings in general are much more powerful and complex than at first look. Once you understand them well they are an
+extremely powerful features. When use correctly they give you an expressive way to communicate with your
+users and dependees as well as you in the future. Misused they can be just noise that trains everybody to ignore them. 
+
+Here is a quick tip/summary of what this section will expand upon. 
+
+ - Always set `warnings.warn(stacklevel=...)` to the right value (at least 2). 
+ - Be descriptive of what "deprecated" means. 
+ - Be descriptive of what the replacement is.
+ - Always indicate since when it is deprected / the replacement is available.
+ - Don't be afraid to use multiple line strings.
+
+
+## Always use `stacklevel=...`
+
+
+Setting the stacklevel ensure that python reports the right place where the deprecated feature is used. 
+
+### Make it easier to fix
+
+```
+# file example.py
+import warnings
+def function(argument=None):
+   if argument:
+      warnings.warn('`argument is deprecated`')
+```
+
+Will lead to the following error:
+
+```
+~/example.py:5: UserWarning: `argument is deprecated`
+  warnings.warn('`argument is deprecated`')
+```
+
+While setting `stacklevel=2` will point to the right file, line, and show the problematic code:
+
+```
+~/foo.py:3: UserWarning: `argument is deprecated`
+  example.function(1)
+```
+
+Most terminal and editor now recognize the `file/path.py:LineNumber` syntax, so on my setup it is also a single click to
+open the right file on the right location and fix it. 
+
+### makes it easier to find
+
+Most test runners have options to turns DeprecationWarnings into error **only in the code you maintain**. 
```

Comment:
I think turning `DeprecationWarnings` warnings into errors for your libraries dependencies is also very valuable. (For example, when NumPy is issuing deprecation warnings and your library depends on NumPy)

Reply:
Ah but what I am trying to convey here is the same, if numpy `stacklevel` is set properly turrning all DeprecationWarnigns in napari code into errors, will let you know that you need to update your code, while still letting numpy use deprecated functionalities internally.